### ISSUE TITLE
Type cleaning

### ISSUE THREAD

**Issue by albsch** [etylizer/etylizer#71]
It is not clear when type cleaning as implemented in #59 can be applied during type checking. 

With the new inference branch merged (#27), the cleaning after a substitution is applied from tally generates errors for simple tests.

**Comment by skogsbaer** [comment]
We can do the cleaning right after applying a substitution, but we can only clean type variables
that were freshly introduced by tally. Here is the relevant quote from the paper
"Polymorphic Functions with Set-Theoretic Types Part 2" by Castagna et al:

> Finally, solve introduces several fresh polymorphic variables which can be cleaned up after that the substitutions have been applied to obtain the types deduced by inference system: all variables that occur only in covariant (resp. contravariant) position in a type, can be replaced by 0 (resp. 1).

In some email thread, I argued that the set of type variables freshly introduced by tally is **different** from the set of fixed type variables given to tally.

There are three classes of type variables: universally quantified type variables, type variables freshly introduced by constraint generation and simplification, and type variables freshly introduced by tally. I now investigate which type variables must be fixed and which can be cleaned.

## Universally quantified type variables.

Consider

```erlang
-spec foo(integer()) -> integer().
foo(X) -> X.

-spec bar(T, T) -> integer().
bar(X, _Y) -> foo(X).
```

If tally would be allowed to substitute `T`, then it would unify `T` with `integer()` and the example would type check. Clearly, this is wrong: tally must not touch `T`, so `T` is a fixed type variable. It's also obvious that such type variables must not be cleaned.

##  Type variables freshly introduced by constraint generation and simplification

Tally is allowed to substitute such type variables (indeed, that's the whole point of invoking tally in the first place!).
But we must not clean such type variables because the encode incomplete knowledge.

Here is an example that uses type inference:

```erlang
case_01(X) ->
    case X of
        spam -> foobar;
        _ -> foobar
    end.
```

We assign `X` some fresh type variable `$1`. Constraint generation then assumes that the
srutiny of `case X of ...` has type `$3`, requiring `$1 <= $3`. Further the "guard type" `t` for the
case `spam -> foobar` is `$3 /\ spam`.

During constraint simplifiation, tally solves the constraint set `{$1 <= $3}` with the
substitution `sigma = {$1 => mu13 /\ $3}`. Further, it checks for unreachable branches
by checking whether `t sigma <= bottom` (where `t` is the guard type of the branch).

`t sigma` (the application of substitution `sigma` to `t`) yields `t`. But if we allowed type
variables freshly introduced by constraint generation to be cleaned, cleaning `t = $3 /\ spam`
would yield `bottom`. Hence, the branch `spam -> foobar` would be reported as unreachable. This
is clearly wrong.

Here is a similar example without inference:

```erlang
-spec foo() -> foobar.
foo() ->
  F = fun(X) ->
        case X of
            spam -> foobar;
            _ -> foobar
        end
      end,
  F(spam).
```

## Type variables freshly introduced by tally

These can't fixed (because they are not in the input to tally) but subjec to cleaning.

## Summary

* Universally quantified type variables: fixed, no cleaning
* Type variables freshly introduced by constraint generation and simplification: not fixed, no cleaning
* Type variables freshly introduced by tally: cleaning

## Implementation

@albsch said that it requires some work to get the freshly introduced type variables out
of tally. So here's a simple implementation idea:

- Rewrite `clean` so that it takes a predicate stating which type variables should be cleaned
(instead of just a set of type variables).
- Before applying tally, compute the set of free type variables of the constraint set. Then
the predicate for `clean` just checks whether a type variable is in this set or not.

**Comment by skogsbaer** [comment]
@albsch Does the above sound good to you?

**Comment by albsch** [comment]
First let me shortly explain what tally (the invocation $tally_\Delta(C)$ ) is doing. The algorithm knows two types of variables: 

* A variable $\alpha$ is fixed (or monomorphic), if it is contained in the fixed variable set $\Delta$: $\alpha \in \Delta$
* A variable $\alpha$ is not fixed (or polymorphic), if it is not contained in the fixed variable set $\Delta$: $\alpha \not \in \Delta$

Tally is invoked on a set of constraints $C$ and given such a set of fixed variables $\Delta$. Let us assume tally will tag freshly introduced variables with a generated tag. Let's fix some terminology on variables:

* $free(C)$: All variables in $C$ not contained in $\Delta$
* $fixed(C)$: All variables in $C$ contained in $\Delta$
* $generated(C)$: All variables in $C$ generated and tagged by tally and not in $\Delta$
* $nongenerated(C)$: All variables in $C$ not touched tally and not in $\Delta$
* $top(C)$: All variables in $C$ on the top level and not contained in $\Delta$

These definitions are naturally extended if instead of a constraint set $C$ the functions are given a set of constraint sets $S$ or a set $\Sigma$ of substitutions $\sigma$. The invocation of tally on a constraint set $C$ returns such a set of solution substitutions $\Sigma$ for example.

It should be noted that $clean_\Delta(\Sigma)$ also operates only on fixed vs non-fixed variables. In the case of CDuce, the $\Delta$ given to clean is the same $\Delta$ given to the invocation of $tally_\Delta(C)$, as cleaning is applied directly after tally.

Some invariants will be maintained:

* $generated(C) \cup nongenerated(C) = free(C)$
* $fixed(tally_\Delta(C)) \subseteq fixed(C)$
  * Tally does not touch the fixed variables, does not add any, but fixed variables can disappear, if constraints are redundant
* $free(C) \cap \Delta = \emptyset$
* $free(tally_\Delta(C)) \cap \Delta = \emptyset$
  * A free variable (both generated and non-generated) will never become a fixed variable, they will always be different, even after a tally invocation

> I argued that the set of type variables freshly introduced by tally is different from the set of fixed type variables given to tally.

This is indeed the case, as tally will adhere to the constraints given by $\Delta$.

Universally quantified type variables
===

The consequence is that `T` must be supplied to tally and clean in the $\Delta$ set. Tally will therefore not touch `T`.

Type variables freshly introduced by constraint generation and simplification
===

> During constraint simplifiation, tally solves the constraint set {$1 <= $3} with the
substitution sigma = {$1 => mu13 /\ $3}

This is not the correct tally solution. If `$1` and `$3` are not in $\Delta$ (and are both top-level variables, i.e. not hidden below a type constructor), tally *will* touch these variables and generate fresh variables out of those. 
A correct solution---depending on variable ordering---for $tally_\emptyset(\\$1 <= \\$3)$ is:

* $\\$3 => mu3$
* $\\$1 => mu1 \cap mu3$

In general, the non-generated set is always empty, if all variables are contained on the top level:
* $top(C) = free(C) \rightarrow nongenerated(tally_\Delta(C)) = \emptyset$.

Cleaning `t = $mu3 /\ spam` would indeed yield bottom, because `mu3` was introduced by tally.

If both $1 and $3 are given in the fixed set, then the solution of $tally_\emptyset(\\$1 <= \\$3)$ is unsatisfiable, as both $1 and $3 are treated as monomorphic variables.

Type variables freshly introduced by tally
===

These are variables in the $generated$ set. Personally, I would assume that after invocation of tally, this equation holds: $free(C) = generated(tally_\Delta(C))$ and $nongenerated(tally_\Delta(C)) = \emptyset$, i.e. tally touches **all** variables not contained in $\Delta$. In general, tally touches all variables that are not hidden below a type constructor.

Summary
===

> not fixed, no cleaning

Tally and cleaning cannot do this. Either some variables are fixed, and they will not be cleaned. Or, some variables are not fixed, and they will be cleaned. As in the solution for the constraint $1 < $3 with variables $1 and $3 example, it's impossible to do both. Either both $1 and $3 are in $\Delta$, or not.

**Comment by skogsbaer** [comment]
@albsch Thanks for your careful analysis.

Two comments: 

Firstly, you state that {$1 => mu13 /\ $3} is not tally's solution to {$1 <= $3}. But that's what I got by just pasting the output from the log. What's wrong here?

Seondly and more importantly: both type variables $1 and $3 are not fixed and they should not be cleaned because they already existed before invoking tally.

There seems to be some misunderstanding. From the point of cleaning, it doesn't matter what tally does internally. The only thing that matters is the following. Assume $S = tally_{\Delta}(C)$ where $C$ is a set of constraints and $S$ is a set of substitutions. Assume further $\sigma \in S$. Then we can clean type variable $\alpha$ if it is free in $\sigma$ but not free in $C$.

That's what I implemented in #87. All tests are working (the failures come from erlang 26) and bugs #36 and #56 are also fixed.

**Comment by albsch** [comment]
A valid tally implementation adheres to the following completeness theorem:

![image](https://github.com/etylizer/etylizer/assets/38429047/26ddaa2e-3f4b-4d4f-81fb-53047f0380c6)

The point I'm trying to make is the following. Whether `$3` is substituted by tally or not is up to the implementation. According to the completeness theorem it's still complete.

> Secondly and more importantly: both type variables $1 and $3 are not fixed and they should not be cleaned because they already existed before invoking tally.

`$1` and `$3` will not exist anymore after the tally invocation if they are not in the fixed set $\Delta$. They will be substituted to a fresh variable `$mu1` and `$mu3`. That this is not the case in the implementation is a detail that I have skipped because I think it is not important, but I think I should do it to make it clear that tally is allowed to substitute **every** variable that is not contained in $\Delta$.

> Then we can clean type variable $\alpha$ if it is free in $\sigma$ but not free in $\Delta$.

This applied to **every** variable in $\sigma$ after a tally invocation (that is not in $\Delta$ and adheres to the co-contravariance position restriction). Again, every variable that is not in $\Delta$ will be fresh after a tally invocation. That this is not the case is only because I have not implemented traversing the solution and substituting the remaining variables. Tally is allowed to do that because of the completeness theorem.

**Comment by skogsbaer** [comment]
Let me put the other way round: we must not clean type variables free in C. Otherwise, the examples `case_01` and `case_01_noinfer` from https://github.com/etylizer/etylizer/issues/71#issuecomment-1895313502 fail to type check.

On the other hand, we usually do not fix all type variables in C. Hence, there exists type variables that are not fixed and which must not be cleaned.


Can you investigate what happens with examples `case_01` and `case_01_noinfer` if tally would substitute all variables with fresh variables? You would have to work in branch #87 or you need to do the calculations by hand with the information from the log.

**Comment by albsch** [comment]
I implemented the tally fix, in `as/clean-after-subst-vars`.

`case_01` and most tests are ok.

Test case `case_01_infer` is one of the tests where the new tally mechanism generates fresh variables for the solution.
In our case, the solution is

`{$2 := mu$45 /\ mu$23, $4 := mu$45}`

whereas before it was

`{$2 := $4 /\ mu$23, $4 := $4}`

Both solutions are valid solutions which satisfy the input constraints (sanity condition passes).

Test case:

```erlang
-spec case_01_noinfer() -> foobar.
case_01_noinfer() ->
  F = fun(X) -> case X of spam -> foobar; _ -> foobar end end,
  F(spam).
```

The test case fails with:

```
[INFO 2024-01-24 10:28:41.989 typing.erl:255] Checking failed: test_files/tycheck_simple.erl:176:27: Type error: this branch never matches
%  176|   F = fun(X) -> case X of spam -> foobar; _ -> foobar end end,
%     |                           ^

  while checking function case_01_noinfer/0 against type fun(() -> foobar)
test_files/tycheck_simple.erl:176:27: Type error: this branch never matches
%  176|   F = fun(X) -> case X of spam -> foobar; _ -> foobar end end,
%     |    

while checking function case_01_noinfer/0 against type fun(() -> foobar)
```

**Comment by skogsbaer** [comment]
@albsch got you. Currently, I don't how enough time to think about it carefully. Two points: 

- I think that my approach to detect the type variables that can be cleaned (outlined above and implemented in #87) is wrong.
- I still think that tally should return a set with type variables that can be cleaned, together with the substitution.

Maybe the following from the part 2 paper is relevant: 

> Finally, **solve** introduces several fresh polymorphic variables which can be cleaned up after that the substitutions have been applied to obtain the types deduced by inference system.

It does not state that the type variables freshly introduced by **tally** can be cleaned but only those introduced by **solve**.

@albsch does this make sense? Can you research in this direction?

**Comment by albsch** [comment]
> It does not state that the type variables freshly introduced by tally can be cleaned but only those introduced by solve.

Solving introduces these fresh variables and is the last phase of tally. It is also the only place where fresh variables are introduced.

> I still think that tally should return a set with type variables that can be cleaned, together with the substitution.

Tally can return that set. With the implementation in `as/clean-after-subst-vars` it will always return the set where every variable that is contained in $free(C)$ is replaced with a (universally quantified, unbounded) fresh variable. 

I think we should try to understand why different valid implementations of tally (w.r.t. the completeness theorem) lead to different behavior under cleaning.

**Comment by skogsbaer** [comment]
The paper only introduces cleaning with respect to the concrete algorithm presented. Cleaning is not stated as a general thing that can be applied to every implementation obeying to the soundness and completeness properties of tally.

What's the output of the algorithm from the paper when applied to the constraint set `{$1 <= $3}`?

**Comment by skogsbaer** [comment]
@albsch Do we need this issue anymore? #36 and #56 are fixed with PR #91.
